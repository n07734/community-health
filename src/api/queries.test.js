import {
    prQuery,
    reviewCommentsQuery,
    commentsQuery,
    reviewsQuery,
    releasesQuery,
    issuesQuery,
} from './queries'

describe('queries:', () => {
    const makeRawData = ({ type = 'repository', resultType }) => {
        return {
            data: {
                [type]: {
                    id: 'ID',
                    [resultType]: {
                        edges: [1,2,3],
                        pageInfo: {
                            hasNextPage: true,
                            endCursor: 'endCursor',
                        },
                    },
                },
            },
        }
    }

    const makeStateArgs = (args) => ({
        fetches: {
            org: 'org',
            repo: 'repo',
            ...args,
        },
    })


    const testArgs = [
        {
            name: 'issuesQuery',
            testFunction: issuesQuery,
            stateArgs: {
                issuesPagination: {
                    cursor: 'cursor',
                },
            },
            rawDataArgs: {
                type: 'repository',
                resultType: 'issues',
            },
        },
        {
            name: 'releasesQuery',
            testFunction: releasesQuery,
            stateArgs: {
                releasesPagination: {
                    cursor: 'cursor',
                },
            },
            rawDataArgs: {
                type: 'repository',
                resultType: 'releases',
            },
        },
        {
            name: 'reviewsQuery',
            testFunction: reviewsQuery,
            queryArgs: {
                nodeId: 'nodeId',
                cursor: 'cursor',
            },
            rawDataArgs: {
                type: 'node',
                resultType: 'reviews',
            },
            expectFillerType: 'pullRequestReviewComments',
        },
        {
            name: 'commentsQuery',
            testFunction: commentsQuery,
            queryArgs: {
                nodeId: 'nodeId',
                cursor: 'cursor',
            },
            rawDataArgs: {
                type: 'node',
                resultType: 'comments',
            },
        },
        {
            name: 'reviewCommentsQuery',
            testFunction: reviewCommentsQuery,
            queryArgs: {
                nodeId: 'nodeId',
                cursor: 'cursor',
            },
            rawDataArgs: {
                type: 'node',
                resultType: 'comments',
            },
        },
        {
            name: 'prQuery',
            testFunction: prQuery,
            stateArgs: {
                prPagination: {
                    cursor: 'cursor',
                },
            },
            rawDataArgs: {
                type: 'repository',
                resultType: 'pullRequests',
            },
            expectFillerType: 'pullRequests',
        },
    ]

    testArgs
        .forEach(({ name, testFunction, stateArgs, queryArgs, rawDataArgs, expectFillerType }) => {
            describe(`${name}:`, () => {
                const args = stateArgs
                    ? makeStateArgs(stateArgs)
                    : queryArgs

                const {
                    query,
                    resultInfo,
                    fillerType,
                } = testFunction(args)

                it('query to match snapshot', () => {
                    expect(query).toMatchSnapshot()
                })

                it('fillerType to be corretct', () => {
                    expect(fillerType).toEqual(expectFillerType)
                })

                it('resultInfo', () => {
                    const data = makeRawData(rawDataArgs)

                    const {
                        rawData,
                        results,
                        hasNextPage,
                        nextArgs,
                    } = resultInfo(data)

                    expect(rawData).toEqual(data)
                    expect(results).toEqual([1,2,3])
                    expect(hasNextPage).toEqual(true)
                    expect(nextArgs).toEqual({
                        ...(
                            rawDataArgs.type !== 'node'
                                &&  {
                                    org: 'org',
                                    repo: 'repo',
                                }
                        ),
                        nodeId: 'ID',
                        cursor: 'endCursor',
                    })
                })
            })
        })
})
