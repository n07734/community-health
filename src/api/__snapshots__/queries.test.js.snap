// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`queries: commentsQuery: query to match snapshot 1`] = `
"{
      node(id:\\"nodeId\\") {
          ... on PullRequest {
            id
            
    comments(first: 100  after:\\"cursor\\" ) {
      edges {
        node {
          author {
            login
            url
          }
          body
        }
      }
      pageInfo { endCursor hasNextPage hasPreviousPage startCursor }
    }

          }
        }
    }"
`;

exports[`queries: reviewCommentsQuery: query to match snapshot 1`] = `
"{
      node(id:\\"nodeId\\") {
          ... on PullRequestReview {
            id
            
    comments(first: 100  after:\\"cursor\\" ) {
      edges {
        node {
          author {
            login
            url
          }
          body
        }
      }
      pageInfo { endCursor hasNextPage hasPreviousPage startCursor }
    }

          }
        }
    }"
`;

exports[`queries: reviewsQuery: query to match snapshot 1`] = `
"{
      node(id:\\"nodeId\\") {
          ... on PullRequest {
            id
            
    reviews(first: 100  after:\\"cursor\\" ) {
      edges {
        node {
          id
          state
          author {
            login
            url
          }
          
    comments(first: 10 ) {
      edges {
        node {
          author {
            login
            url
          }
          body
        }
      }
      pageInfo { endCursor hasNextPage hasPreviousPage startCursor }
    }

        }
      }
      pageInfo { endCursor hasNextPage hasPreviousPage startCursor }
    }

          }
        }
    }"
`;
